test/framework/machinehealthcheck_helpers.go

```diff
@@ -0,0 +1,194 @@
+type GetMachineHealthChecksByClusterInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+}
+
+// GetMachineHealthChecksByCluster returns the MachineHealthCheck objects for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachineHealthChecksByCluster(ctx context.Context, input GetMachineHealthChecksByClusterInput) []*clusterv1.MachineHealthCheck {
+	machineHealthCheckList := &clusterv1.MachineHealthCheckList{}
+	Expect(input.Lister.List(ctx, machineHealthCheckList, byClusterOptions(input.ClusterName, input.Namespace)...)).To(Succeed(), "Failed to list MachineDeployments object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	machineHealthChecks := make([]*clusterv1.MachineHealthCheck, len(machineHealthCheckList.Items))
+	for i := range machineHealthCheckList.Items {
+		machineHealthChecks[i] = &machineHealthCheckList.Items[i]
+	}
+	return machineHealthChecks
+}
+
+// machineHealthCheckOptions returns a set of ListOptions that allows to get all machine objects belonging to a MachineHealthCheck.
+func machineHealthCheckOptions(machineHealthCheck clusterv1.MachineHealthCheck) []client.ListOption {
+	return []client.ListOption{
+		client.MatchingLabels(machineHealthCheck.Spec.Selector.MatchLabels),
+	}
+}
+
+// WaitForMachineHealthCheckToRemediateUnhealthyNodeConditionInput is the input for WaitForMachineHealthCheckToRemediateUnhealthyNodeCondition.
+type WaitForMachineHealthCheckToRemediateUnhealthyNodeConditionInput struct {
+	ClusterProxy        ClusterProxy
+	Cluster             *clusterv1.Cluster
+	MachineHealthChecks []*clusterv1.MachineHealthCheck
+}
+
+// WaitForMachineHealthCheckToRemediateUnhealthyNodeCondition patches a node condition to any one of the machines with a node ref.
+func WaitForMachineHealthCheckToRemediateUnhealthyNodeCondition(ctx context.Context, input WaitForMachineHealthCheckToRemediateUnhealthyNodeConditionInput, intervals ...interface{}) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for PatchNodeConditions")
+	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling PatchNodeConditions")
+	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling PatchNodeConditions")
+
+	for _, mhc := range input.MachineHealthChecks {
+		mhc := mhc
+		fmt.Fprintf(GinkgoWriter, "Waiting until the node with unhealthy node condition is remediated\n")
+		Eventually(func() bool {
+			machines := GetMachinesByMachineHealthCheck(context.TODO(), GetMachinesByMachineHealthCheckInput{
+				Lister:             input.ClusterProxy.GetClient(),
+				ClusterName:        input.Cluster.Name,
+				Namespace:          input.Cluster.Namespace,
+				MachineHealthCheck: mhc,
+			})
+			if len(machines) == 0 {
+				return true
+			}
```

Comment:
This seems like it will give us a false positive. When will there be no matching machines?

Reply:
you are right, this shouldn't be valid.